### PR TITLE
Replace Input Manager with Input Action

### PR DESCRIPTION
**Pull request by iamteapot422**
Gathers all controls into the input action asset, creates InputSystem. Most of Input.mousePosition weren't replaced, because it takes a lot more code with Mouse.current.position.

### CODE REVIEW

File: Assets/Scenes/Startup.unity [hunk 1]
```diff
@@ -1274,33 +1274,14 @@ GameObject:
   m_Component:
```
Comment (stilnat) [inline Assets/Scenes/Startup.unity]: Can you describe what changes you did to the startup scene ?

File: Assets/Scripts/SS3D/Networking/NetworkPhysicsDebug.cs [hunk 2]
```diff
@@ -18,8 +19,14 @@ public class NetworkPhysicsDebug : NetworkActor
 
         public List<Rigidbody> SpawnedCans;
 
+        protected override void OnStart()
+        {
+            base.OnStart();
+            SystemLocator.Get<InputSystem>().Inputs.Other.SpawnCans.performed += SpawnSodaCans;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/NetworkPhysicsDebug.cs]: rename SpawnSodaCans in HandleCansSpawned

File: Assets/Scripts/SS3D/Systems/Entities/Humanoid/HumanoidController.cs [hunk 3]
```diff
@@ -54,10 +62,31 @@ protected override void OnStart()
...
         }
 
+        protected void OnEnable()
+        {
+            try
+            {
+                MovementControls.Enable();
+                HotkeysControls.Enable();
+            }
+            catch (Exception) { }
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Entities/Humanoid/HumanoidController.cs]: what's the use of this try catch block here ?

File: Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs [hunk 4]
```diff
@@ -44,25 +46,10 @@ protected override void OnStart()
...
-            if (Input.GetKeyDown(KeyCode.Tab))
-            {
-                _fade.SetFade(true);
-            }
-            if (Input.GetKeyUp(KeyCode.Tab))
-            {
-                _fade.SetFade(false);
-            }
+            Controls.OtherActions controls = SystemLocator.Get<InputSystem>().Inputs.Other;
+            controls.Fade.performed += _ => _fade.SetFade(true);
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs]: I would rather avoid lambda expressions in this case, makes things hard to debug if needed later on. Tell me if you disagree. Otherwise replace with good old private methods HandleFadePerformed and HandleFadeCancelled.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 5]
```diff
@@ -33,66 +35,71 @@ public class ConsolePanelView : Actor
...
-            ProcessInput();
             if (_isSliding)
             {
                 Slide();
             }
         }
-    
-        private void ProcessInput()
+
+        private void HandleClose(InputAction.CallbackContext context)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:78]: Add a summary explaining what this does.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 6]
```diff
@@ -33,66 +35,71 @@ public class ConsolePanelView : Actor
...
+            _consoleControls.Disable();
+            _consoleControls.Open.Enable();
+        }
 
-            if ((GetArrowVertical() != 0) && _isShowed)
-            {
-                _chosenPrevCommand = Math.Clamp(_chosenPrevCommand - GetArrowVertical(), 0, _allPrevCommands.Count - 1);
-                _inputField.text = _allPrevCommands[_chosenPrevCommand];
-            }
+        private void HandleOpen(InputAction.CallbackContext context)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:92]: Add a summary explaining what this does.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 7]
```diff
@@ -33,66 +35,71 @@ public class ConsolePanelView : Actor
...
+            _targetPointMin = new Vector2(0, -_consolePanel.rect.height);
+            _targetPointMax = _targetPointMin + new Vector2(0, _consolePanel.rect.height);
+            _inputField.Select();
+            _controls.Disable();
+            _consoleControls.Enable();
+            _consoleControls.Open.Disable();
         }
 
-        private int GetArrowVertical()
+        private void HandleSwitchCommand(InputAction.CallbackContext context)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:105]: Add a summary explaining what this does.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 8]
```diff
@@ -33,66 +35,71 @@ public class ConsolePanelView : Actor
...
-            {
-                return -1;
-            }
-            return 0;
+            _chosenPrevCommand =
+                Math.Clamp(_chosenPrevCommand + (int)context.ReadValue<float>(), 0, _allPrevCommands.Count - 1);
+            _inputField.text = _allPrevCommands[_chosenPrevCommand];
         }
 
+        private void HandleSubmit(InputAction.CallbackContext context)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:114]: Add a summary explaining what this does.

File: Assets/Scripts/SS3D/Systems/Input/Controls.cs [hunk 9]
```diff
@@ -0,0 +1,1025 @@
+//------------------------------------------------------------------------------
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Input/Controls.cs:1]: I'm going to need some help understanding this one.

File: Assets/Scripts/SS3D/Systems/Input/Controls.inputactions [hunk 10]
```diff
@@ -0,0 +1,587 @@
+{
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Input/Controls.inputactions]: I'm going to need help with understanding this one too.

File: Assets/Scripts/SS3D/Systems/Input/InputSystem.cs [hunk 11]
```diff
@@ -0,0 +1,14 @@
+using SS3D.Core;
+using UnityEngine;
+
+public class InputSystem : SS3D.Core.Behaviours.System
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Input/InputSystem.cs:6]: Add a summary explaining what this does.

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs [hunk 12]
```diff
@@ -34,42 +35,20 @@ public override void OnStartClient()
...
-            {
-                ProcessSecondaryClick();
-            }
-
-            if (Input.GetButtonDown("Use"))
-            {
-                ProcessUse();
-            }
+            Controls controls = SystemLocator.Get<InputSystem>().Inputs;
+            controls.Other.PrimaryClick.performed += HandlePrimaryClick;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs]: I'm concerned with the absence of unsuscribing from events overall. Is there a particular reason for that ?

File: Assets/Scripts/SS3D/Systems/Screens/GameScreensController.cs [hunk 13]
```diff
@@ -21,6 +21,14 @@ protected override void OnStart()
             _menuOpen = true;
             _blockNone = true;
             _spawnedState = PlayerSpawnedState.IsNotSpawned;
+            SystemLocator.Get<InputSystem>().Inputs.Other.Menu.performed += _ =>
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Screens/GameScreensController.cs]: same as previous comment, I'd raver have a good old method here, but maybe that's just me. What's the reason here for the lambda expression ? Is there any other reason than having the method directly next to the suscribing ?

File: Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs [hunk 14]
```diff
@@ -114,19 +123,21 @@ public override void Update()
             {
                 return;
             }
-
-            // Hand-related buttons
-            if (Input.GetButtonDown("Swap Hands") && HandContainers.Length > 0)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs]: why remove the HandContainers > 0 check ?

File: Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs [hunk 15]
```diff
@@ -152,7 +163,8 @@ public void SetActiveHand(AttachedContainer selectedContainer)
                 Debug.LogError("selectedContainer is not in HandContainers.");
             }
         }
-
+        
+        // This method can't be a callback because of Fishnet issue
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs]: Not sure I understand this comment, please explain it to me.

**Review by stilnat — CHANGES_REQUESTED** [review]
Overall, the few changes needed are : 
- Unsuscribe from events that you previously suscribed too.
- Add some comments where I asked, and maybe more if you feel fancy.
- Discuss with me about the lambda expressions.
- We'll also need some documentation regarding this new InputSystem, since it must become the new norm to get input. If you agree, I'll add you as editor on gitbook and you write a little something about it (mainly how to use it in code, how to add new controls and stuff.)

Great work !

File: Assets/Scenes/Startup.unity (hunk 1 above)
Comment (iamteapot422) [reply Assets/Scenes/Startup.unity]: Added InputSystem.cs to the camera and set ScrollView speed in Console to 32

File: Assets/Scripts/SS3D/Systems/Entities/Humanoid/HumanoidController.cs (hunk 3 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Entities/Humanoid/HumanoidController.cs]: OnEnable() is being called before OnStart(), so Controls are still empty. I have found try catch is the only solution to enable Controls in this situation, but forgot to specify Exception.

File: Assets/Scripts/SS3D/Systems/Input/Controls.cs (hunk 9 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Input/Controls.cs:1]: This file is auto-generated class of Controls.inputactions, it changes every time you save changes in .inputacitons

File: Assets/Scripts/SS3D/Systems/Input/Controls.inputactions (hunk 10 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Input/Controls.inputactions]: This is where the input settings are. You can change control maps/action maps/actions and their bindings/interactions/processors through the editor

File: Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs (hunk 14 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs]: It's moved to HandleSwapHands

File: Assets/Scripts/SS3D/Systems/Input/InputSystem.cs (hunk 11 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Input/InputSystem.cs:6]: It's just contains one single Controls instance. Is summary really necessary?

File: Assets/Scripts/SS3D/Networking/NetworkPhysicsDebug.cs [hunk 16]
```diff
@@ -18,15 +19,31 @@ public class NetworkPhysicsDebug : NetworkActor
         public float ForceMultiplier = 1f;
 
         public List<Rigidbody> SpawnedCans;
+        private Controls.OtherActions _controls;
+
+        private void OnEnable()
+        {
+            try
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Networking/NetworkPhysicsDebug.cs]: what the try catch block for is here ?

File: Assets/Scripts/SS3D/Systems/Entities/Debug/MindSwapDebug.cs [hunk 17]
```diff
@@ -9,20 +10,33 @@ public class MindSwapDebug : NetworkActor
     {
         public Entity Origin;
         public Entity Target;
+        private Controls.OtherActions _controls;
+
+        private void OnEnable()
+        {
+            try
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Entities/Debug/MindSwapDebug.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/Entities/Humanoid/HumanoidController.cs [hunk 18]
```diff
@@ -53,41 +52,40 @@ public abstract class HumanoidController : NetworkActor
...
         protected void OnEnable()
         {
             try
             {
-                MovementControls.Enable();
-                HotkeysControls.Enable();
+                MovementControls.ToggleRun.performed += HandleToggleRun;
             }
-            catch (Exception) { }
+            catch (NullReferenceException) {}
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Entities/Humanoid/HumanoidController.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs [hunk 19]
```diff
@@ -19,9 +18,26 @@ public class GamemodeObjectivePanelView : Actor
...
         [SerializeField] private GamemodeObjectiveItemView _itemViewPrefab;
         [SerializeField] private GameObject _content;
+        private Controls.OtherActions _controls;
 
         private Dictionary<int, GamemodeObjectiveItemView> _gamemodeObjectiveItems;
 
+        private void OnEnable()
+        {
+            try
+            {
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Gamemodes/UI/GamemodeObjectivePanelView.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 20]
```diff
@@ -23,8 +22,8 @@ public class ConsolePanelView : Actor
         [SerializeField] private float _movingSpeed = 2500f;
         [SerializeField] private TMP_InputField _inputField;
         // Used for opening/closing
-        private bool _isSliding = false;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:24]: why the = false removed ?

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 21]
```diff
@@ -34,23 +33,35 @@ public class ConsolePanelView : Actor
         private CommandsController _commandsController;
         // Used for choosing command via arrows
         [SerializeField] private List<string> _allPrevCommands = new() {""};
-        private int _chosenPrevCommand = 0;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:35]: why the =0 removed ?

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 22]
```diff
@@ -34,23 +33,35 @@ public class ConsolePanelView : Actor
...
+        private int _chosenPrevCommand;
         private Controls _controls;
         private Controls.ConsoleActions _consoleControls;
 
+        private void OnEnable()
+        {
+            try
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/Input/InputSystem.cs (hunk 11 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Input/InputSystem.cs:6]: It's a bit verbose, I agree with you, but we might auto generate some doc in the future, and having a summary on each public classes even the trivial ones would be good for that.

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs [hunk 23]
```diff
@@ -59,10 +77,6 @@ private void HandleSecondaryClick(InputAction.CallbackContext callbackContext)
         [Client]
         private void HandleUse(InputAction.CallbackContext callbackContext)
         {
-            if (!enabled)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs]: why is this removed ?

File: Assets/Scripts/SS3D/Systems/Interactions/RadialInteractionView.cs [hunk 24]
```diff
@@ -39,14 +39,28 @@ public sealed class RadialInteractionView : Core.Behaviours.System
 
         private List<IInteraction> Interactions { get; set; }
         private InteractionEvent Event { get; set; }
+        private Controls.OtherActions _controls;
+
+        private void OnEnable()
+        {
+            try
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/RadialInteractionView.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/Interactions/RadialInteractionView.cs [hunk 25]
```diff
@@ -173,10 +188,6 @@ private void Show()
 
         private void HandleDisappear(InputAction.CallbackContext callbackContext)
         {
-            if (!enabled)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Interactions/RadialInteractionView.cs]: why is this removed ?

File: Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs [hunk 26]
```diff
@@ -81,11 +81,32 @@ public class CameraFollow : Actor
         private const float SnapAngle = 45.1f;
 
         #endregion
-        
+
+        private void OnEnable()
+        {
+            try
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs [hunk 27]
```diff
@@ -99,10 +120,6 @@ protected override void HandleUpdate(in float deltaTime)
         }
         private void HandleZoom(InputAction.CallbackContext context)
        {
-           if (!enabled)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs]: why is this removed ?

File: Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs [hunk 28]
```diff
@@ -116,11 +133,7 @@ private void HandleSnapRight(InputAction.CallbackContext context)
        }
         private void Snap(float direction)
         {
-            if (!enabled)
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs]: why is this removed ?

File: Assets/Scripts/SS3D/Systems/Screens/GameScreensController.cs [hunk 29]
```diff
@@ -13,6 +15,21 @@ public sealed class GameScreensController : NetworkActor
         [SerializeField] private bool _blockNone;
         [SerializeField] private bool _menuOpen;
         private PlayerSpawnedState _spawnedState;
+        private Controls.OtherActions _controls;
+
+        private void OnEnable()
+        {
+            try
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Screens/GameScreensController.cs]: replace with` if(null && !didstart) return; else { Debug.Error(); return; }` So we can still have an error in case performed is null later in the game.

File: Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs [hunk 30]
```diff
@@ -52,6 +52,19 @@ public class Hands : InteractionSource, IToolHolder, IInteractionRangeLimit, IIn
 
         public HandsView HandsView { get; private set; }
 
+        private void OnEnable()
+        {
+            _controls = SystemLocator.Get<InputSystem>().Inputs.Hotkeys;
```
Comment (stilnat) [inline Assets/Scripts/SS3D/Systems/Storage/Containers/Hands.cs]: doesn't this need a null check too ?

**Review by stilnat — CHANGES_REQUESTED** [review]
Some conversation resolved, some added. Most are duplicated and can be solved in a single answer.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs (hunk 20 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:24]: Redundant

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs (hunk 21 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs:35]: Redundant

File: Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs (hunk 23 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Interactions/InteractionController.cs]: Redundant

File: Assets/Scripts/SS3D/Systems/Interactions/RadialInteractionView.cs (hunk 25 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Interactions/RadialInteractionView.cs]: Redundant

File: Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs (hunk 27 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs]: Redundant

File: Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs (hunk 28 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Screens/CameraFollow.cs]: Redundant

**Review by stilnat — DISMISSED** [review]
Approved on my side, don't merge yet, waiting for John's input regarding the script order thing.
Also have to play test it again.

**Comment by cosmiccoincidence** [comment]
Will need to merge in master again & resolve conflicts due to a lot of recent changes.

File: Assets/Scripts/SS3D/Systems/Input/Controls.inputactions (hunk 10 above)
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Input/Controls.inputactions]: This file should not be in the Scripts folder. Could be added to `Content/Input`
image

File: Assets/Scripts/SS3D/Systems/Input/InputSystem.cs [hunk 31]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public Controls Inputs { get; private set; }
+    protected override void OnAwake()
+    {
+        base.OnAwake();
+        
+        Inputs = new Controls();
+        Inputs.Other.Enable();
+    }
+}
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Input/InputSystem.cs]: The whole InputSystem stuff can be moved to an assembly. It does not depend on any other assembly for that,

Comment (joaoburatto) [reply Assets/Scripts/SS3D/Systems/Input/InputSystem.cs]: And apply proper spacing on that, those methods are touching each other. No touching.

File: Assets/Scripts/SS3D/Systems/Screens/GameScreensController.cs [hunk 32]
```diff
@@ -13,6 +15,18 @@ public sealed class GameScreensController : NetworkActor
         [SerializeField] private bool _blockNone;
         [SerializeField] private bool _menuOpen;
         private PlayerSpawnedState _spawnedState;
+        private Controls.OtherActions _controls;
+
+        private void OnEnable()
+        {
+            _controls = SystemLocator.Get<InputSystem>().Inputs.Other; 
+            _controls.Menu.performed += HandleMenuPerformed;
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/Screens/GameScreensController.cs]: HandleMenuPerformed? What does Menu.performed even means? This is not a clear name for this action.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs [hunk 33]
```diff
@@ -32,67 +33,93 @@ public class ConsolePanelView : Actor
...
+            _consoleControls.Submit.performed += HandleSubmit;
+        }
+
+        private void OnDisable()
+        {
+            _consoleControls.Close.performed -= HandleClose;
+            _consoleControls.Open.performed -= HandleOpen;
+            _consoleControls.SwitchCommand.performed -= HandleSwitchCommand;
+            _consoleControls.Submit.performed -= HandleSubmit;
+        }
```
Comment (joaoburatto) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs]: If this was the issue solved by changing the script order, simply move to OnStart and OnDestroy

**Review by joaoburatto — DISMISSED** [review]

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs (hunk 33 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/IngameConsoleSystem/ConsolePanelView.cs]: Move what?

**Review by stilnat — APPROVED** [review]
ready to go
